app/presenters/pages_presenter.rb

```diff
@@ -0,0 +1,124 @@
+    @items = pagination.paginate(pages)
+    @url_params = url_params
+    @sort_key = sort_key
+    @sort_dir = sort_dir
+    @start_date = start_date
+    @end_date = end_date
+  end
+
+  def rows
+    items.map do |page|
+      [
+        { text: page_text(page[:base_path]) },
+        { text: page[:survey_count], format: "numeric" },
+        { text: top_phrases(page) },
+      ]
+    end
+  end
+
+  def head
+    [
+      page_title_head,
+      surveys_answered_head,
+      {
+        text: "Most frequent phrases",
+      },
+    ]
+  end
+
+private
+
+  attr_reader :pagination, :url_params, :sort_key, :sort_dir, :start_date, :end_date
+  attr_accessor :page_titles
+
+  def page_text(base_path)
+    title = page_title(base_path)
+    "<a class='display-block' href='base_path'>#{title}</a><span class='govuk-body-s display-block'>#{base_path}</span>".html_safe # rubocop:disable Rails/OutputSafety
+  end
+
+  def page_title(base_path)
+    # Search API doesn't index all documents, so it won't always return
+    # a result for every base path we pass it.
+    return search_api_page_titles[base_path] if search_api_page_titles.has_key?(base_path)
+
+    # If the search results don't include the base path we want
+    # we hit the Content Store API, which will
+    content_store_page_title(base_path)
+  end
+
+  def search_api_page_titles
+    @search_api_page_titles ||= begin
+      Services.search_api.search(
+        count: items.count,
+        filter_link: item_base_paths,
+        fields: %i[link title],
+      )
+        .to_hash["results"]
+        .each_with_object({}) { |result, hash| hash[result["link"]] = result["title"] }
+    end
+  end
+
+  def item_base_paths
+    @item_base_paths ||= items.map { |item| item[:base_path] }
+  end
+
+  def content_store_page_title(base_path)
+    Services.content_store.content_item(base_path)["title"]
+  end
+
+  def page_title_head
+    key = "page_base_path"
+    {
+      text: "Page title",
+      href: href(key),
+    }.merge(sort_direction(key))
+  end
+
+  def surveys_answered_head
+    key = "feedback_comments"
+    {
+      text: "Surveys answered",
+      format: "numeric",
+      href: href(key),
+    }.merge(sort_direction(key))
+  end
+
+  def href(key)
+    direction_for_link = sort_dir
+    if currently_sorting_by_key?(key)
+      direction_for_link = opposite_sort_dir
+    end
+
+    sort_params = { sort_key: key, sort_dir: direction_for_link }
+    pages_path(url_params.merge(sort_params))
+  end
+
+  def opposite_sort_dir
+    sort_dir == "desc" ? "asc" : "desc"
+  end
+
+  def sort_direction(key)
+    return {} unless currently_sorting_by_key?(key)
+
+    { sort_direction: sort_dir }
+  end
+
+  def currently_sorting_by_key?(key)
+    key == sort_key
+  end
+
+  def top_phrases(page)
+    Page.top_phrases(page[:page_id], start_date, end_date)
+      .take(3)
+      .map { |phrase| "<span class='phrase'>#{phrase}</span>" }
```

Comment:
Should this instead be passed in from the controller?

Reply:
Oh, good shout, removes need for passing in start/end date

Reply:
Oh heck. The pages haven't been paginated in the controller so when it iterates over all of them to get the top phrases, there are thousands of pages so it takes 67,000ms to render the page! Moving the pagination to out of the presenter is a bit eggy, and duplicating the logic around pagination in the controller is undesirable. I agree with your thought that it should be done in the controller but there doesn't seem to be a good way of doing this that I can think of right now. If its ok, I'm going to move it back to making this query in the presenter and we can revisit it when we know more

Reply:
Ok, let's come back to this as soon as we can.